Automattic/newspack-plugin
# Google Ad Manager Feature Extraction

### All Submissions:






### Changes proposed in this Pull Request:

This PR describes the extraction of the Google Ad Manager functionality to a feature plugin. To keep things simple I'll describe the complete change here even though it technically spans three repositories. 

This work establishes a pattern for extracting Newspack functionality to feature plugins. The approach:

- Move all new functionality to a feature plugin (`newspack-{feature-description}`)
- Create a Configuration Manager class to interact with the plugin.
- Try to limit all Newspack Plugin Wizard activity to API endpoints.

The elements of this PR are:

- Created a feature plugin (https://github.com/Automattic/newspack-gam) to contain Newspack Google Ad Manager features.
- Moved the Ad Manager block from https://github.com/Automattic/newspack-blocks/pull/39 to the new plugin. When this lands, https://github.com/Automattic/newspack-blocks/pull/39 can be abandoned and closed.
- Moved all of the Ad Unit model code from `newspack-plugin` to `newspack-gam` (https://github.com/Automattic/newspack-gam/blob/master/includes/class-newspack-gam-model.php), including [unit tests](https://github.com/Automattic/newspack-gam/tree/master/tests/unit-tests). The `newspack-plugin` [Wizard's PHP class](https://github.com/Automattic/newspack-plugin/compare/update/use-gam-plugin?expand=1#diff-7037d54febdb3a09e7453fc8eb4516d7) is dramatically simpler, containing only API endpoints and script/style enqueueing. 
- Created a [Configuration Manager class](https://github.com/Automattic/newspack-plugin/compare/update/use-gam-plugin?expand=1#diff-0af9a1b981f2f2343689b108042cf35f) for `newspack-gam`.  All interaction with the feature plugin occurs in this class. At first glance this class may seem unnecessary, but I like adhering to the Wizard->Configuration Manager->Plugin model even when the plugin is our own.
- The [Google Ad Manager block](https://github.com/Automattic/newspack-gam/tree/master/src/blocks/google-ad-manager) addresses feedback from https://github.com/Automattic/newspack-gam/issues/1. 
- https://github.com/Automattic/newspack-plugin/blob/master/includes/wizards/class-google-ad-manager-wizard.php#L141-L150 was removed. If the AMP plugin detects an `amp-ad` element, it will automatically add this script.
- Declared  `newspack-gam` as a required plugin for the Ad Manager Wizard. 

### How to test the changes in this Pull Request:

1. Apply this PR to the Newspack plugin and run `npm run build`
1. Clone https://github.com/Automattic/newspack-gam in your test environment's plugins directory. Run `composer install`. Run `npm install && npm run build:webpack`. After this branch lands, I'll generate a release of `newspack-gam` and add the path to the .zip to `newspack-plugin`.
1. Navigate to the Ad Manager Wizard. This should install `newspack-gam` before opening.
1. Create/Edit/Delete ad units. Sample AMP code can be found in the description of https://github.com/Automattic/newspack-blocks/pull/39
1. Create a post and add the Google Ad Manager block (found in the Newspack category).
1. Verify the dropdown lists all Ad Units you have set up.
1. If you have set up no Ad Units verify the block shows UI with text and a link to create new Ad Units.
1. Verify the block placeholder dimensions change to match the Ad Unit you have selected.
1. Verify that choosing an Ad Unit that does not contain proper dimension data displays a block error notification. Also verify that switching to a valid Ad Unit clears the notification.
1. Verify that all unit tests pass in `newspack-gam`
1. Disable `Newspack Blocks` and verify that the `Google Ad Manager` block is still available, in the Common category.  

### Other information:
